src/ts/background/handlers/State.ts

```diff
@@ -0,0 +1,217 @@
+import {
+  AuthorizeRequest,
+  MessageAuthorize,
+  MessageExtrinsicSign,
+  MessageExtrinsicSign$Response,
+  SigningRequest
+} from '../types'
+
+import extension from 'extensionizer'
+import { BehaviorSubject } from 'rxjs'
+import { assert } from '@polkadot/util'
+
+type AuthRequest = {
+  id: string,
+  idStr: string,
+  request: MessageAuthorize,
+  resolve: (result: boolean) => void,
+  reject: (error: Error) => void,
+  url: string
+}
+
+type AuthUrls = {
+  [index: string]: {
+    count: number,
+    id: string,
+    isAllowed: boolean,
+    origin: string,
+    url: string
+  }
+}
+
+type SignRequest = {
+  id: string,
+  request: MessageExtrinsicSign,
+  resolve: (result: MessageExtrinsicSign$Response) => void,
+  reject: (error: Error) => void,
+  url: string
+}
+
+let idCounter = 0
+
+function getId (): string {
+  return `${Date.now()}.${++idCounter}`
+}
+
+export default class State {
+  // at the moment, we are keeping the list in memory - this should be persisted
+  private _authUrls: AuthUrls = {}
+  private _authRequests: { [index: string]: AuthRequest } = {}
+  private _signRequests: { [index: string]: SignRequest } = {}
+  private _windows: Array<number> = []
+  readonly authSubject: BehaviorSubject<Array<AuthorizeRequest>> =
+    new BehaviorSubject([] as Array<AuthorizeRequest>)
+  readonly signSubject: BehaviorSubject<Array<SigningRequest>> =
+    new BehaviorSubject([] as Array<SigningRequest>)
+
+  get numAuthRequests (): number {
+    return Object.keys(this._authRequests).length
+  }
+
+  get numSignRequests (): number {
+    return Object.keys(this._signRequests).length
+  }
+
+  get allAuthRequests (): Array<AuthorizeRequest> {
+    return Object
+      .values(this._authRequests)
+      .map(({ id, request, url }) => [id, request, url])
+  }
+
+  get allSignRequests (): Array<SigningRequest> {
+    return Object
+      .values(this._signRequests)
+      .map(({ id, request, url }) => [id, request, url])
+  }
+
+  private popupClose (): void {
+    this._windows.map((id: number) =>
+      extension.windows.remove(id)
+    )
+    this._windows = []
+  }
+
+  private popupOpen (height, width): void {
+    extension.windows.create({
+      focused: true,
+      height: height,
+      left: Math.floor((window.screen.availWidth - width) / 2),
+      top: Math.floor((window.screen.availHeight - height) / 2),
+      type: 'popup',
+      url: extension.extension.getURL('popup.html'),
+      width: width
+    }, (window?: chrome.windows.Window) => {
```

Comment:
Does this work on firefox as it refers to `chrome.windows.Window`?

Reply:
Good catch again. Replaced chrome.windows with polyfill Object Windows